File: linked-lists/js/src/doubly-linked-list.test.js
import { describe, it } from "jsr:@std/testing/bdd";
import { expect } from "jsr:@std/expect";
import { DoublyLinkedList } from "./doubly-linked-list.js";

describe("append", () => {
  it("should append a node to the end of an empty list", () => {
    // Arrange.
    const list = new DoublyLinkedList();

    // Act.
    const actual = list.append("first");

    // Assert.
    expect(actual.value).toEqual("first");
    expect(list.head.value).toEqual("first");
  });

  it("should append a node to the end of a list", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node = {
      value: "first",
      next: null,
    };
    list.head = node;
    list.tail = node;

    // Act.
    const actual = list.append("second");

    // Assert.
    expect(actual.value).toEqual("second");
    expect(list.head.value).toEqual("first");
    expect(list.head.next.value).toEqual("second");
    expect(list.tail.prev.value).toEqual("first");
  });

  it("should append a node to the end of a longer list", () => {
    // Arrange.
    const list = new DoublyLinkedList();

    // Act.
    list.append("first");
    list.append("second");
    list.append("third");
    const actual = list.append("fourth");

    // Assert.
    expect(actual.value).toEqual("fourth");
    expect(list.head.value).toEqual("first");
    expect(list.head.next.value).toEqual("second");
    expect(list.tail.value).toEqual("fourth");
    expect(list.tail.prev.value).toEqual("third");
  });
});

describe("insertAt", () => {
  it("should insert a node after the node passed", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node = {
      value: "first",
      next: null,
    };
    list.head = node;

    // Act.
    const actual = list.insertAt(node, "second");

    // Assert.
    expect(actual.value).toEqual("second");
    expect(list.head.value).toEqual("first");
    expect(list.head.next.value).toEqual("second");
  });
});

describe("remove", () => {
  it("should do nothing when removing a node from an empty list", () => {
    // Arrange.
    const list = new DoublyLinkedList();

    // Act.
    const actual = list.remove("value");

    // Assert.
    expect(actual).toBeUndefined();
    expect(list.head).toBeNull();
  });

  it("should remove a node from a list of one", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node = {
      value: "first",
      next: null,
    };
    list.head = node;

    // Act.
    list.remove(node);

    // Assert.
    expect(list.head).toBeNull();
  });

  it("should remove the last node from a list of three nodes", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node2 = {
      value: "second",
    };
    const node3 = {
      value: "third",
      next: null,
      prev: node2,
    };
    list.tail = node3;
    node2.next = node3;
    const node1 = {
      value: "first",
      next: node2,
    };
    node2.prev = node1;
    list.head = node1;

    // Act.
    list.remove(node3);

    // Assert.
    expect(list.head.value).toEqual("first");
    expect(list.head.next.value).toEqual("second");
    expect(list.tail.value).toEqual("second");
    expect(list.tail.prev.value).toEqual("first");
  });

  it("should remove the first node from a list of three nodes", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node2 = {
      value: "second",
    };
    const node3 = {
      value: "third",
      next: null,
      prev: node2,
    };
    list.tail = node3;
    node2.next = node3;
    const node1 = {
      value: "first",
      next: node2,
    };
    node2.prev = node1;
    list.head = node1;

    // Act.
    list.remove(node1);

    // Assert.
    expect(list.head.value).toEqual("second");
    expect(list.head.next.value).toEqual("third");
    expect(list.tail.value).toEqual("third");
    expect(list.tail.prev.value).toEqual("second");
  });

  it("should remove a node from the middle of a list of three nodes", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node2 = {
      value: "second",
    };
    const node3 = {
      value: "third",
      next: null,
      prev: node2,
    };
    list.tail = node3;
    node2.next = node3;
    const node1 = {
      value: "first",
      next: node2,
    };
    node2.prev = node1;
    list.head = node1;

    // Act.
    list.remove(node2);

    // Assert.
    expect(list.head.value).toEqual("first");
    expect(list.head.next.value).toEqual("third");
    expect(list.tail.value).toEqual("third");
    expect(list.tail.prev.value).toEqual("first");
  });
});

describe("traverse", () => {
  it("should return null when traversing an empty list", () => {
    // Arrange.
    const list = new DoublyLinkedList();

    // Act.
    const actual = list.traverse((node) => node.value === "second");

    // Assert.
    expect(actual).toBeNull();
  });

  it("should traverse the list and find the node requested", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node2 = {
      value: "second",
    };
    const node3 = {
      value: "third",
      next: null,
      prev: node2,
    };
    list.tail = node3;
    node2.next = node3;
    const node1 = {
      value: "first",
      next: node2,
    };
    node2.prev = node1;
    list.head = node1;

    // Act.
    const actual = list.traverse((node) => node.value === "second");

    // Assert.
    expect(actual.value).toEqual("second");
  });

  it("should traverse the list to the end", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node2 = {
      value: "second",
    };
    const node3 = {
      value: "third",
      next: null,
      prev: node2,
    };
    list.tail = node3;
    node2.next = node3;
    const node1 = {
      value: "first",
      next: node2,
    };
    node2.prev = node1;
    list.head = node1;

    // Act.
    const actual = list.traverse((_) => false);

    // Assert.
    expect(actual.value).toEqual("third");
  });
});

describe("traverseBackward", () => {
  it("should return null when traversing an empty list", () => {
    // Arrange.
    const list = new DoublyLinkedList();

    // Act.
    const actual = list.traverseBackward((node) => node.value === "second");

    // Assert.
    expect(actual).toBeNull();
  });

  it("should traverse the list backward and find the node requested", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node2 = {
      value: "second",
    };
    const node3 = {
      value: "third",
      next: null,
      prev: node2,
    };
    list.tail = node3;
    node2.next = node3;
    const node1 = {
      value: "first",
      next: node2,
    };
    node2.prev = node1;
    list.head = node1;

    // Act.
    const actual = list.traverseBackward((node) => node.value === "second");

    // Assert.
    expect(actual.value).toEqual("second");
  });

  it("should traverse to the beginning of the list", () => {
    // Arrange.
    const list = new DoublyLinkedList();
    const node2 = {
      value: "second",
    };
    const node3 = {
      value: "third",
      next: null,
      prev: node2,
    };
    list.tail = node3;
    node2.next = node3;
    const node1 = {
      value: "first",
      next: node2,
    };
    node2.prev = node1;
    list.head = node1;

    // Act.
    const actual = list.traverseBackward((_) => false);

    // Assert.
    expect(actual.value).toEqual("first");
  });
});

File: linked-lists/go/pkg/double/doubly-linked-list_test.go
package double

import "testing"

func TestAppend(t *testing.T) {
	type want struct {
		val  string
		head string
	}
	tests := []struct {
		name      string
		emptyList bool
		value     string
		head      string
		want      want
	}{
		{
			"should append a node to the end of an empty list",
			false,
			"first",
			"first",
			want{"first", "first"},
		},
		{
			"should append a node to the end of a list",
			false,
			"first",
			"first",
			want{"first", "first"},
		},
	}

	t.Parallel()

	for _, tc := range tests {
		tt := tc
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			ll, _ := genList(tt.emptyList)

			got := ll.Append(tt.value)

			// TODO: add in .prev checks
			if tt.want.val != got.Val {
				t.Errorf("want = %s, got = %s", tt.want.val, got.Val)
			}
			if tt.want.head != ll.head.Val {
				t.Errorf("want = %s, got = %s", tt.want.head, ll.head.Val)
			}
		})
	}
}

func TestInsertAt(t *testing.T) {
	type want struct {
		head    string
		val     string
		nextVal string
	}
	tests := []struct {
		name      string
		emptyList bool
		value     string
		insertAt  string
		want      want
	}{
		{
			"should insert the node passed",
			false,
			"third",
			"first",
			want{"first", "third", "second"},
		},
	}

	t.Parallel()

	for _, tc := range tests {
		tt := tc
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			ll, nodes := genList(tt.emptyList)

			got := ll.InsertAt(nodes[tt.insertAt], tt.value)

			if tt.want.val != got.Val {
				t.Errorf("want = %s, got = %s", tt.want.val, got.Val)
			}
			if tt.want.nextVal != got.next.Val {
				t.Errorf("want = %s, got = %s", tt.want.nextVal, got.next.Val)
			}
			if tt.want.head != ll.head.Val {
				t.Errorf("want = %s, got = %s", tt.want.head, ll.head.Val)
			}
		})
	}
}

func TestRemove(t *testing.T) {
	type want struct {
		head        *Node
		nextVal     string
		nextNextVal string
	}
	tests := []struct {
		name      string
		emptyList bool
		remove    string
		want      want
	}{
		{
			"should do nothing when removing node from empty list",
			true,
			"third",
			want{},
		},
		{
			"should remove node from list",
			false,
			"third",
			want{&Node{Val: "first"}, "second", "fourth"},
		},
	}

	t.Parallel()

	for _, tc := range tests {
		tt := tc
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			ll, nodes := genList(tt.emptyList)

			ll.Remove(nodes[tt.remove])

			if tt.want.head != nil {
				if tt.want.head.Val != ll.head.Val {
					t.Errorf("want = %s, got = %s", tt.want.head.Val, ll.head.Val)
				}
				if tt.want.nextVal != ll.head.next.Val {
					t.Errorf("want = %s, got = %s", tt.want.nextVal, ll.head.next.Val)
				}
				if tt.want.nextNextVal != ll.head.next.next.Val {
					t.Errorf("want = %s, got = %s", tt.want.nextNextVal, ll.head.next.next.Val)
				}
			}
		})
	}
}

func TestTraverse(t *testing.T) {
	type want struct {
		val      string
		head     string
		nextNode *Node
	}
	tests := []struct {
		name        string
		emptyList   bool
		predicateFn func(*Node, *Node) bool
		want        want
	}{
		{
			"should do nothing in an empty list",
			true,
			func(n, _ *Node) bool { return n.Val == "third" },
			want{},
		},
		{
			"should traverse a list to the end",
			false,
			func(_, _ *Node) bool { return false },
			want{"fourth", "first", nil},
		},
		{
			"should traverse a list and find the node requested",
			false,
			func(n, _ *Node) bool { return n.Val == "third" },
			want{"third", "first", &Node{Val: "fourth"}},
		},
	}

	t.Parallel()

	for _, tc := range tests {
		tt := tc
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			ll, _ := genList(tt.emptyList)

			got := ll.Traverse(tt.predicateFn)

			if (want{}) == tt.want && ll.head == nil {
				return
			}

			if tt.want.head != ll.head.Val {
				t.Errorf("want = %s, got = %s", tt.want.head, ll.head.Val)
			}
			if tt.want.val != got.Val {
				t.Errorf("want = %s, got = %s", tt.want.val, got.Val)
			}
			if tt.want.nextNode != nil && tt.want.nextNode.Val != got.next.Val {
				t.Errorf("want = %s, got = %s", tt.want.nextNode.Val, got.next.Val)
			}
		})
	}
}

func TestTraverseBackward(t *testing.T) {
	tests := []struct {
		name        string
		emptyList   bool
		predicateFn func(*Node, *Node) bool
		want        *Node
		nextNode    *Node
	}{
		{
			"should do nothing in an empty list",
			true,
			func(n, _ *Node) bool { return n.Val == "third" },
			nil,
			nil,
		},
		{
			"should traverse a list to the beginning",
			false,
			func(_, _ *Node) bool { return false },
			&Node{Val: "first"},
			&Node{Val: "second"},
		},
		{
			"should traverse a list backward and find the node requested",
			false,
			func(n, _ *Node) bool { return n.Val == "third" },
			&Node{Val: "third"},
			&Node{Val: "fourth"},
		},
	}

	t.Parallel()

	for _, tc := range tests {
		tt := tc
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			ll, _ := genList(tt.emptyList)

			got := ll.TraverseBackward(tt.predicateFn)

			if tt.want == nil && tt.want != got {
				t.Errorf("want = %s, got = %s", tt.want.Val, got.Val)
				return
			}
			if tt.want != nil && tt.want.Val != got.Val {
				t.Errorf("want = %s, got = %s", tt.want.Val, got.Val)
			}
			if tt.want != nil && tt.nextNode.Val != got.next.Val {
				t.Errorf("want = %s, got = %s", tt.want.Val, got.next.Val)
			}
		})
	}
}

func genList(empty bool) (DoublyLinkedList, map[string]*Node) {
	if empty {
		return DoublyLinkedList{}, map[string]*Node{}
	}

	node4 := Node{
		Val:  "fourth",
		next: nil,
	}
	node3 := Node{
		Val:  "third",
		next: &node4,
	}
	node2 := Node{
		Val:  "second",
		next: &node3,
	}
	node1 := Node{
		Val:  "first",
		next: &node2,
		prev: nil,
	}
	node2.prev = &node1
	node3.prev = &node2
	node4.prev = &node3
	nodes := map[string]*Node{
		"first":  &node1,
		"second": &node2,
		"third":  &node3,
		"fourth": &node4,
	}
	ll := DoublyLinkedList{
		head: &node1,
		tail: &node4,
	}

	return ll, nodes
}